### ISSUE TITLE
Cannot set limits on SQLite connections ( like SQLITE_MAX_ATTACHED )

### ISSUE THREAD

**Issue by BastouP411** [xerial/sqlite-jdbc#523]
I want to prevent users from attaching other databases, but I am not able to set any limit, is there any way of doing it ?

**Comment by BastouP411** [comment]
Solved with https://github.com/xerial/sqlite-jdbc/pull/527#issue-462402792

**Comment by jehugaleahsa** [comment]
My question about this enhancement is whether it will allow setting MAX_ATTACH_LIMIT to something higher (which is a compile time setting) or only allow them to be decreased?

Sorry I don't understand how the SQLite piece gets compiled, but it'd be really nice if the compilation took place with all of the maximums in place. For example, with MAX_ATTACH_LIMIT the maximum is 125 but the default is 10. It'd be nice if SQLite was compiled with 125 but then the connection defaulted it to 10 unless I pass in the `setPragma` explicitly. I hope that made sense.

Hoping to see this feature soon. I recently stumbled on an interesting use case where I need more than the max attach limit and I'd love to see the updated maven artifact. 👍

**Comment by BastouP411** [comment]
> My question about this enhancement is whether it will allow setting MAX_ATTACH_LIMIT to something higher (which is a compile time setting) or only allow them to be decreased?
> 
> Sorry I don't understand how the SQLite piece gets compiled, but it'd be really nice if the compilation took place with all of the maximums in place. For example, with MAX_ATTACH_LIMIT the maximum is 125 but the default is 10. It'd be nice if SQLite was compiled with 125 but then the connection defaulted it to 10 unless I pass in the `setPragma` explicitly. I hope that made sense.
> 
> Hoping to see this feature soon. I recently stumbled on an interesting use case where I need more than the max attach limit and I'd love to see the updated maven artifact. 👍

The modification that I made uses a native function from the SQLite C library. You should be able to set any limit to any value with the pragma.

For example  `jdbc:sqlite:sample.db?limit_attached=125` should work and give you a maximum of 125 attached databases.

**Comment by jehugaleahsa** [comment]
Ah, okay. When I was reading the documentation for https://www.sqlite.org/c3ref/limit.html I got the impression it could only be used to reduce the limit, not increase it, for ATTACH anyway. This is really exciting then!

**Comment by BastouP411** [comment]
The maximum for `limit_attached` is still 125 but you can make it higher by using the pragma

**Comment by BastouP411** [comment]
The pull request was merged. I’mclosing the issue.

**Comment by jehugaleahsa** [comment]
So good and bad news. I pulled the SNAPSHOT version and stepped through it. I confirmed that I can decrease the ATTACH limit from 10 down to something smaller:
![image](https://user-images.githubusercontent.com/1368126/90203017-7b8adc00-ddad-11ea-891c-40f057e0877a.png)

However, if I try to go past 10, the limit stays at 10. I re-read the SQLite docs again and they do make it pretty clear you can only _decrease_ the limit at runtime. In order to increase it, you'd need to update your compilation step to override the default limits: https://www.sqlite.org/c3ref/c_limit_attached.html

I was curious how compilation took place so I took a closer look over the weekend and I see now that various docker images are created, the compiled output is stuck in the JAR and they all get shipped together. In order to increase these limits, you'd need to update the MAKE file to say `-DSQLITE_LIMIT_ATTACHED=125`, but then you'd also need to lower it to 10 via these changes to match the recommended defaults, as most people wouldn't want these maxed out, by default, since a lot of these limits are for security reasons.

**Comment by xerial** [comment]
Ok. So, the next steps are:
- Recompile native binaries with -DSQLITE_LIMIT_ATTACHED=(more than 5 (default?)) 
- Setting the default value to 5 when initializing JDBC driver. If a limit configuration is provided, we should use this value.

**Comment by jehugaleahsa** [comment]
It looks like it should be `SQLITE_MAX_ATTACHED` actually. I was looking through the C header this morning trying to figure this out and `SQLITE_LIMIT_` is just a "category" declaration but I guess the flags starting with `SQLITE_MAX_` are picked up by a preprocessor to apply these changes. https://www.sqlite.org/limits.html#max_attached They couldn't have made this much harder to find information on.

**Comment by JaegerOmega** [comment]
Serious new person question... So if I wanted to set this to a max of say 50, how would one go about doing it/building it?

**Comment by jehugaleahsa** [comment]
@JaegerOmega I was hoping to work on this enhancement a few weeks ago and figure out the build process, but after spending an hour getting `make` to work with Mac newlines, I threw my hands up. Hopefully someone with a little more patience or experience with the project comes along.

I was following along with the `README_BUILD.md`: https://github.com/xerial/sqlite-jdbc/blob/master/README_BUILD.md. It seems to work, but I had to keep replacing newlines on my windows machine and use the windows subsystem for linux (ubuntu with build-essentials and perl installed). I did this for about an hour before I started to cry. I am pretty sure you just need to tack the new limit macros onto the end of this file here: https://github.com/xerial/sqlite-jdbc/blob/master/Makefile#L86. I think then it is just taking the changes in this PR (calling `sqlite3_limit`) to set the current defaults in the constructor, unless overridden with `setPragma`. Seems totally achievable.

**Comment by jehugaleahsa** [comment]
@JaegerOmega I had some more time this morning to play around with setting this environment up for development. It turns out all of my issues were caused by me using Git for Windows to clone the repo and then trying to use wsl (Windows Subsystem for Linux) to develop. After deleting my copy and using `git clone` inside a wsl sessions, `make` worked as expected.

The problem is Git for Windows converts `\n` to `\r\n` which confused the Linux tools when they encountered the additional `\r`. Seems obvious in retrospect, but I only figured it out when I opened VS Code and it said CRLF at the bottom of the file. Previously I was thinking this project stored files using Mac newlines but I was wrong. So I can make progress again...

The next step on this journey is following through the somewhat [long-winded process](https://nickjanetakis.com/blog/setting-up-docker-for-windows-and-wsl-to-work-flawlessly) of getting docker to run in wsl/Ubuntu. Note the `README_BUILD.md` suggests using Cygwin for Windows development, but I want to continue down this path using wsl since it seems like best practices these days. If I can create a successful build using wsl, I will probably include a separate PR just to update the build steps for Windows. This all seems to work well with VS Code's [WSL - Remote](https://code.visualstudio.com/docs/remote/remote-overview) extension so far.

I hope I am really close, then I can finally focus on trying out the changes I'd like to make regarding increasing these limits. 😅
